### PR TITLE
feat: add option to ignore initContainers volumes in validation

### LINKED ISSUE #82: Feature Request: Add option to ignore volumes mounted by Init Containers
### Is your feature request related to a problem?

about: Propose a new feature to enhance the policy's flexibility
title: 'Feature: Add option to ignore volumes mounted by Init Containers'
labels: 'feature, enhancement'

### Is your feature request related to a problem? Please describe.

Currently, the `volumes-psp-policy` validates all volumes defined within a Pod's `spec.volumes` against the `allowedTypes` list. This includes volumes that are mounted by `initContainers`. In some scenarios, `initContainers` might require specific volume types (e.g., `hostPath` for bootstrapping or temporary storage) that are generally not desired for regular application containers. This can lead to policy rejections for valid use cases of `initContainers` if their required volume types are not in the `allowedTypes` list, forcing users to either relax the policy for all containers or create more complex workarounds.

### Solution you'd like

_No response_

### Alternatives you've considered

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by vvlisn**
## Description

Adds a new policy setting: ignoreInitContainersVolumes.
When enabled, volumes that are only used by initContainers (and not by main containers) are ignored during validation.

Fix #82 

Test
To test this pull request, you can run the following commands:
Apply to Makefile
bats
Unit tests and e2e tests have been updated to cover the new setting and its edge cases.
You can also manually test with:
Apply to Makefile
'
Additional Information
Tradeoff
The new setting is optional and defaults to false, so it does not affect existing users.
Only minimal code changes were made to maximize backward compatibility and ease of review.
Potential improvement
Future: Support for skipping validation for volumes used by other container types (e.g., ephemeralContainers) if needed.
More granular control (e.g., per-volume or per-container) could be considered if there is demand.

### CODE REVIEW

File: validate.go [hunk 1]
```diff
@@ -59,15 +84,23 @@ func validate(payload []byte) ([]byte, error) {
...
 		})
 
+		if settings.IgnoreInitContainersVolumes {
+			// Skip volumes that are only used by initContainers and not by containers
+			if _, found := initContainerVolumeNames[volumeName]; found {
+				if _, usedByContainer := containerVolumeNames[volumeName]; !usedByContainer {
+					continue
+				}
+			}
+		}
```
Comment (flavio) [inline validate.go:94]: An "invalid" volume used both by an init container and by one of the main containers will always be considered valid. I don't like that.

File: validate.go [hunk 2]
```diff
@@ -59,15 +84,23 @@ func validate(payload []byte) ([]byte, error) {
...
+
 		if !settings.AllowedTypes.Contains(volumeType) {
 			errMsg := fmt.Sprintf("volume '%s' of type '%s' is not in the AllowedTypes list",
 				volumeName, volumeType)
 			if err == nil {
-				err = errors.New(errMsg)
+				err = fmt.Errorf("%s", errMsg)
```
Comment (flavio) [inline validate.go:100]: why are you doing this change?

**Review by flavio — CHANGES_REQUESTED** [review]
Can you also update the README to document this new setting and its behavior?

**Review by jvanz — CHANGES_REQUESTED** [review]

File: validate.go (hunk 1 above)
Comment (vvlisn) [reply validate.go:94]: > The current implementation does not consider a volume valid just because it is used by an init container. If a volume is used by both an init container and a main container, and its type is not allowed, the request will still be rejected. Only volumes that are exclusively used by init containers are skipped when ignoreInitContainersVolumes is enabled.
>
> Example:
> - If a volume is used by both an init container and a main container, it will always be validated against the allowed types, and the request will be rejected if it is not allowed.

Comment (vvlisn) [reply validate.go:94]: “The current implementation already works as you described: if a volume is used by both initContainers and main containers, and the main container usage is not allowed, the request will be rejected, regardless of the ignoreInitContainersVolumes setting.”

Comment (jvanz) [reply validate.go:95]: Yes, you're right. I missed the `!` in the second if. Sorry for the noise.

File: validate.go [hunk 3]
```diff
@@ -38,6 +38,32 @@ func validate(payload []byte) ([]byte, error) {
...
+		containers := gjson.GetBytes(payload, "request.object.spec.containers")
+		for _, container := range containers.Array() {
+			mounts := container.Get("volumeMounts")
+			for _, mount := range mounts.Array() {
+				name := mount.Get("name").String()
+				if name != "" {
+					containerVolumeNames[name] = struct{}{}
+				}
+			}
+		}
```
Comment (jvanz) [inline validate.go:64]: I think this two for loops could be extracted into a function. I think this will make the code more readable. But I do not consider this a blocker to merge this PR.

**Review by jvanz — APPROVED** [review]

**Comment by vvlisn** [comment]
update README with ignoreInitContainersVolumes setting
refactor: extract volume mount name collection to a helper function"

File: validate.go (hunk 3 above)
Comment (vvlisn) [reply validate.go]: extract volume mount name collection to a helper function"

Comment (jvanz) [reply validate.go]: That's looks great! Thanks for doing it! :heart:

File: validate.go [hunk 4]
```diff
@@ -59,6 +70,16 @@ func validate(payload []byte) ([]byte, error) {
...
 
+		if settings.IgnoreInitContainersVolumes {
+			_, usedByInitContainer := initContainerVolumeNames[volumeName]
+			_, usedByContainer := containerVolumeNames[volumeName]
+
+			if usedByInitContainer && !usedByContainer {
+				// Skip volumes that are only used by initContainers and not by containers
+				continue
+			}
+		}
```
Comment (jvanz) [inline validate.go:81]: :heart:

**Review by flavio — APPROVED** [review]
thanks for the feature and for having implemented the changes we requested

**Comment by flavio** [comment]
today we're going to tag a new release of this policy
